crates/ruff_linter/src/rules/flake8_type_checking/rules/typing_only_runtime_import.rs

```diff
@@ -289,102 +291,116 @@ pub(crate) fn typing_only_runtime_import(
             continue;
         };
 
-        if binding.context.is_runtime()
-            && binding
+        if !binding.context.is_runtime() {
+            continue;
+        }
+
+        let typing_reference = TypingReference::from_references(
+            binding
                 .references()
-                .map(|reference_id| checker.semantic().reference(reference_id))
-                .all(|reference| {
-                    is_typing_reference(reference, &checker.settings().flake8_type_checking)
-                })
-        {
-            let qualified_name = import.qualified_name();
-
-            if is_exempt(
-                &qualified_name.to_string(),
-                &checker
-                    .settings()
-                    .flake8_type_checking
-                    .exempt_modules
-                    .iter()
-                    .map(String::as_str)
-                    .collect::<Vec<_>>(),
-            ) {
-                continue;
-            }
+                .map(|reference_id| checker.semantic().reference(reference_id)),
+            checker.settings(),
+        );
+
+        match typing_reference {
+            TypingReference::No => continue,
+            TypingReference::Future => add_future_import = true,
```

Comment:
Could we avoid doing all the extra work below if the setting is false (I assume the `Yes` and `Quote` cases are the only cases where the rule did something before)
```suggestion
            TypingReference::Future => if allow_future_import { add_future_import = true } else { continue }
```